arbnode/transaction_streamer.go

```diff
@@ -1167,7 +1231,123 @@ func (s *TransactionStreamer) executeMessages(ctx context.Context, ignored struc
 	return s.config().ExecuteMessageLoopDelay
 }
 
+func (s *TransactionStreamer) SubmitEspressoTransactionPos(pos arbutil.MessageIndex) {
+	s.pendingTxnsQueueMutex.Lock()
+	defer s.pendingTxnsQueueMutex.Unlock()
+	s.pendingTxnsPos = append(s.pendingTxnsPos, pos)
+}
+
+func (s *TransactionStreamer) PollSubmittedTransactionForFinality(ctx context.Context) time.Duration {
+
+	data, err := s.espressoClient.FetchTransactionByHash(ctx, s.submittedTxHash)
+	if err != nil {
+		log.Error("failed to fetch the transaction hash", "err", err, "pos", s.submittedTxnPos)
+		return s.config().EspressoTxnsPollingInterval
+	}
+	// get the message at the submitted txn position
+	msg, err := s.getMessageWithMetadataAndBlockHash(*s.submittedTxnPos)
+	if err != nil {
+		log.Error("failed to get espresso message", "err", err)
+		return s.config().EspressoTxnsPollingInterval
+	}
+	// parse the message to get the transaction bytes and the justification
+	txns, jst, err := arbos.ParseEspressoMsg(msg.MessageWithMeta.Message)
+	if err != nil {
+		log.Error("failed to parse espresso message", "err", err)
+		return s.config().EspressoTxnsPollingInterval
+	}
+
+	//If we reach here, the transaction has been included in a block
+	//after confirming the transaction, we need to fetch the header and fill it into the block justification.
+	espressoHeader, err := s.espressoClient.FetchHeaderByHeight(ctx, data.BlockHeight)
+	if err != nil {
+		log.Error("espresso: failed to fetch header by height ", "err", err)
+		return s.config().EspressoTxnsPollingInterval
+	}
+
+	// fetch the namespace proof and vid common. Should use a more efficient way
+	resp, err := s.espressoClient.FetchTransactionsInBlock(ctx, data.BlockHeight, s.config().EspressoNamespace)
+	if err != nil {
+		log.Info("failed to fetch the transactions in block")
+		return s.config().EspressoTxnsPollingInterval
+	}
+
+	// Filling in the block justification with the header
+	jst.Header = espressoHeader
+	jst.Proof = &resp.Proof
+	jst.VidCommon = &resp.VidCommon
+	// create a new message with the header and the txn and the updated block justification
+	newMsg, err := arbos.MessageFromEspressoSovereignTx(txns[0], jst, msg.MessageWithMeta.Message.Header)
+	if err != nil {
+		return s.config().EspressoTxnsPollingInterval
+	}
+	msg.MessageWithMeta.Message = &newMsg
+	batch := s.db.NewBatch()
+	err = s.writeMessage(*s.submittedTxnPos, *msg, batch)
+	if err != nil {
+		return s.config().EspressoTxnsPollingInterval
+	}
+	err = batch.Write()
+	if err != nil {
+		return s.config().EspressoTxnsPollingInterval
+	}
+	s.submittedTxnPos = nil
+	s.submittedTxHash = nil
+	return time.Duration(0)
+}
+
+func (s *TransactionStreamer) submitEspressoTransactions(ctx context.Context, ignored struct{}) time.Duration {
```

Comment:
I think we should emit an info log before or after we submit the transaction to Espresso to make debugging easier.